QR Code generator

adolfogc opened the pull request:
<!--- Provide a general summary of your changes in the title above -->

## Description
This PR is an alternative to PR #1001.

## Motivation and context
This is an improved version of PR #1001 (it also uses libqrencode, instead of qrcodegen).

## How has this been tested?
Manually on macOS.

## Screenshots (if appropriate):
![Screenshot](https://cdn.pbrd.co/images/GT5kLKO.png)

## Types of changes
- ✅ New feature (non-breaking change which adds functionality)
- ✅ Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Please go over all the following points. -->
<!--- Again, remove any lines which don't apply. -->
<!--- Pull Requests that don't fulfill all [REQUIRED] requisites are likely -->
<!--- to be sent back to you for correction or will be rejected.  -->
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- ~☑️ My change requires a change to the documentation and I have updated it accordingly.~
- ~☑️ I have added tests to cover my changes.~

adolfogc commented:
Having finished to organize the GIT history for this PR into a better set of commits, I consider it to be ready for review.

weslly commented:
I'm not sure about the auto-close feature. An user with a bad camera on an old phone may take more than 30 seconds trying to focus a more detailed QR code like the one in the screenshot.

adolfogc commented:
@weslly How much time would be reasonable? 60 seconds?

lenucksi commented:
@adolfogc how about adding a configuration setting for it. I guess 45 seconds should work with most phones, so a default and maybe a few common options would be nice?

weslly commented:
45 seconds is fine for most modern phones but I would prefer not having the auto-close. I think most users would just click the close button before the timer runs out when they are done.

There could also be cases where people open the QR code dialog before getting their phone ready and would have to open it again.

weslly commented:
@adolfogc Maybe we could have something like this instead:
![image](https://user-images.githubusercontent.com/107672/33119330-c4cc5d76-cf55-11e7-80af-715419a13a81.png)

adolfogc commented:
@lenucksi @weslly I'll modify the auto-close feature as follows:
- use 45 seconds by default
- allow the user to disable it by using the checkbox located at the left side of the text (this checkbox will affect the setting globally, not just for the entry currently for which the QR code is shown)

weslly commented:
@adolfogc as long as the user can disable it, you can leave at 30s

adolfogc commented:
@lenucksi @weslly The requested changes have been made. I agree it's more practical to auto-close after 45s, instead of after 30s. The motivation for the auto-close feature is allowing the DB to be locked when KeePassXC is left unattended.

weslly commented on src/gui/MainWindow.ui:
```diff
@@ -545,6 +546,11 @@
   </action>
+  <action name="actionEntryTotpDisplayKey">
+   <property name="text">
+    <string>Show TOTP key</string>
```
Something like "Export TOTP settings" would be better here, since someone could confuse the key as the generated code.

weslly commented:
@adolfogc The app crashes when I try to view the QR code from entries that use the alternative settings format from KeeOtp (`key={secret}&size=6&step=30` at the `otp` attribute)

adolfogc commented:
@weslly Thanks for the review, I'll work on it this weekend.

TheZ3ro commented:
I was looking into this, seems that libqrcodegen it's not in debian repository but I'm not sure. If it is we will have some problem when packaging for debian since we should use debian version of the library (that uses exception)

adolfogc commented:
@TheZ3ro There is no dependency on an external library such as `libqrencode` (this was my previous attempt). All the code that is needed was placed in  the `src/qrcodegen` directory; for QR code generation we're using a stripped down version of Nayuki's QR Code Generator library (qrcodegen), modified to work without raising exceptions.

TheZ3ro commented:
@adolfogc You are including code from another library (qrcodegen) and modifying it for our needs. It's fine from a license point of view, but debian packaging require us to use system library (if present) instead of bundling the library ourselves. 

*My comment was just an heads-up*, currently qrcodegen doesn't seems to be in debian repo, but if in the future it will be included as a separate package we will have some problem.
(I don't understand this clearly, TBH)

Maybe @hifi or @julian-klode can drive us the right way :)

julian-klode commented:
Well, yeah, we don't want embedded code in Debian. If you want to use a library, and want to ship a convience bundle, it should be _unmodified_ so we can link it against a system library. This mostly becomes a problem once multiple programs use the same library and the library gets a security update - it's a question of updating one package vs multiple (and keeping track of these).

Also the COPYING stuff is wrong. You added a GPL-3 license to the files, so they are not MIT licensed, but MIT and (GPL-2 or GPL-3). And while it's legal to add GPLed modifications like that, it's just rude IMO.

hifi commented:
I would probably use a QR code generation library that's well supported across distributions rather than embed if possible.

`qrencode` as previously used seems like something you'd want to link against. Why was it changed to an embedded one?

https://packages.debian.org/stretch/qrencode
https://apps.fedoraproject.org/packages/qrencode

adolfogc commented:
Hi guys, my reply below.

>...but debian packaging require us to use system library (if present) instead of bundling the library ourselves.

>This mostly becomes a problem once multiple programs use the same library and the library gets a security update - it's a question of updating one package vs multiple (and keeping track of these).

Nayuki's QR Code Generator library is relatively new and currently it's distributed as source code (please check its repo here: https://github.com/nayuki/QR-Code-generator). IWhat is the best way to use this library as it is today?

> Also the COPYING stuff is wrong. You added a GPL-3 license to the files, so they are not MIT licensed, but MIT and (GPL-2 or GPL-3). And while it's legal to add GPLed modifications like that, it's just rude IMO.

I don't have much experience with licensing stuff, please understand I wasn't trying to be rude. I just used the same license as the rest of the KeepassXC codebase. Can change it to MIT if it's OK with the KeepassXC developers.

>...qrencode as previously used seems like something you'd want to link against. Why was it changed to an embedded one?

Well, originally I used `libqrencode` but someone pointed to Nayuki's QR code implementation and I though it was a nice alternative. Two reasons for the change were: keeping the sensitive information 
(i.e. keys) inside KeePassXC and not requiring an external dependency.

droidmonkey commented:
The Debian comment mainly related to using distributed libraries to prevent security issues (ie, http).. the qrcode library doesn't fall into that camp.

TheZ3ro commented:
Personally I will prefer a well supported library like qrencode, less code to maintain ourselves in the core codebase

droidmonkey commented on src/qrcodegen/CMakeLists.txt:
```diff
@@ -0,0 +1,25 @@
+#  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+
+include_directories(${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_BINARY_DIR})
```
This definition is global, change to ```target_include_directories``` and move it below add_library if you absolutely need it (you shouldn't).

droidmonkey commented:
I think the use of the qrencode is sound in this case. It is not available as a library, is MIT licensed, and is only a couple source files. There is no real security risk since all operations are conducted locally, and I am kind of on the side of @adolfogc regarding passing secure information to third party libs. You have no idea if they are logging that information insecurely or whatever.

TheZ3ro commented:
If qrcodegen have a security flaw we will need to update it manually, fixing it and release a new keepassxc version.

If we use a shared library, the library will get updated system-wise and you are fine, we(repo-maintainers) just need to compile release with new library version

It's better using a well supported and used library instead af bundling your version that will never be updated.

libqrencode is open source and I think there are no problem passing secure information. (the same thing apply to qrcodegen)

Last, if in the future qrcodegen will be inserted in Debian repository, we will need to refactor this again

phoerious commented on src/gui/TotpKeyDisplayDialog.cpp:
```diff
@@ -0,0 +1,131 @@
+    , m_ui(new Ui::TotpKeyDisplayDialog())
+    , m_secTillClose(AUTOCLOSE_TIME)
+    , m_entry(entry)
+    , m_parent(parent)
```
You don't need to save the parent of a QObject.

phoerious commented on src/gui/TotpKeyDisplayDialog.cpp:
```diff
@@ -0,0 +1,131 @@
+        const QrCode qrc = maybeQRC.value<QrCode>();
+        m_ui->totpKeyLabel->setPixmap(QPixmap::fromImage(qrc.toQImage(200, 4)));
+    } else {
+        auto errorBox = new QMessageBox(m_parent);
```
Use parent(). You can also allocate the message box as an automatic stack variable, so it's immediately deleted when it goes out of scope.

droidmonkey commented:
Well I'm overruled, hah, please also shift back to qrencode library.

phoerious commented:
Just think about it: if we can't get our bundled lib into Debian, we have yet another package we'd need to package for Trusty. 😆

adolfogc commented on src/gui/TotpKeyDisplayDialog.h:
```diff
@@ -0,0 +1,62 @@
+
+public:
+    explicit TotpKeyDisplayDialog(DatabaseWidget* parent = nullptr, Entry* entry = nullptr);
+    ~TotpKeyDisplayDialog();
```
Using `=default` in the class definition caused the generated destructor to be in-lined, which resulted in a compilation error. It turns out you can't use in-line destructors when you use QScopedPointer with a forward declared type (the UI class, in this case). I modified it in the CPP file though.

More info. here:

>**Forward Declared Pointers**
> Classes that are forward declared can be used within QScopedPointer, as long as the destructor of the forward declared class is available whenever a QScopedPointer needs to clean up.
>Concretely, this means that all classes containing a QScopedPointer that points to a forward declared class must have non-inline constructors, destructors and assignment operators: [...]
Source: http://doc.qt.io/qt-5/qscopedpointer.html

phoerious replied:
Don't worry about that then.

adolfogc replied:
I just wanted to justify why I didn't apply this change.

adolfogc commented on src/gui/TotpKeyDisplayDialog.h:
```diff
@@ -0,0 +1,62 @@
+    int m_secTillClose;
+    Entry* m_entry;
+    DatabaseWidget* m_parent;
+    QTimer* m_timer;
```
Not really, I changed them to private.

adolfogc commented on src/gui/DatabaseWidget.cpp:
```diff
@@ -394,6 +395,18 @@ void DatabaseWidget::showTotp()
+{
+    Entry* currentEntry = m_entryView->currentEntry();
+    if (nullptr == currentEntry) {
+        Q_ASSERT(false);
```
In PR #1329 I fixed some other instances of this issue.

phoerious replied:
Cool, will have a look at it later.

hifi commented on src/qrcode/QrCode.cpp:
```diff
@@ -0,0 +1,122 @@
+                                        static_cast<QRecLevel>(ecl));
+}
+
+QVariant QrCode::toQImage(const int size, const int margin) const
```
I've been thinking about this. Qt has a `QSvgGenerator` class that works with `QPainter`. Could you use it instead of `QImage` and make the output to be of some *physical* size taking the screen DPI into account instead of hardcoding the image to 200x200 pixels. The actual vector image can have 1x1 rectangles for each dot but the scaling should be external.

A bonus feature would be that the vector image would scale with the window so you can adjust it.

adolfogc replied:
I don't know if using an SVG would be better. Can you review the changes I've made to address this observation?

hifi replied:
The change to pre-calculate the size before is a definite improvement!

If you're going to keep it as a bitmap, it would make sense to prevent resizing the window as it doesn't serve a useful function.

The change to SVG/vector would hopefully allow the image to automatically resize with the window if properly implemented and you would never need to repaint it manually. That was the main point. Setting the physical size of the bitmap based on DPI is more important than that, though.

adolfogc replied:
> If you're going to keep it as a bitmap, it would make sense to prevent resizing the window as it doesn't serve a useful function.

Agreed.

adolfogc replied:
Although, since QtSVG is already a dependency of KeePassXC, I guess we could try the approach you're proposing and render the SVG using `QGraphicsSvgItem`.

adolfogc replied:
@hifi I will explore the SVG stuff in a next PR if you're OK with it. I don't have more time left in the next couple of days, and they're feature-freezing 2.3.0 soon.

hifi commented on src/gui/TotpExportSettingsDialog.ui:
```diff
@@ -0,0 +1,72 @@
+    <widget class="QLabel" name="totpKeyLabel">
+     <property name="font">
+      <font>
+       <pointsize>53</pointsize>
```
Does the font really need to be of non-standard size? It looks a bit off compared to the rest of the UI.

adolfogc replied:
> Does the font really need to be of non-standard size?

No indeed; this label doesn't render any text though.

phoerious replied:
Generally, do not set font settings inside a UI file. They are absolute and don't scale with the user's DPI configuration.

hifi commented on src/gui/TotpExportSettingsDialog.ui:
```diff
@@ -0,0 +1,72 @@
+       <item>
+        <widget class="QCheckBox" name="enableAutoCloseCheckBox">
+         <property name="text">
+          <string></string>
```
I don't like that this checkbox has two different functions in one and changes text based on checked status. It also moves the checkbox around when it's centered.

adolfogc replied:
Which two functions?

hifi replied:
It affects the current export window and sets a global configure option as well. Also the label should work to change the checked status.

I would prefer removing the auto-close feature completely as suggested before. It's not time critical to close the window and it removes the double function checkbox that moves around.

adolfogc replied:
> It affects the current export window and sets a global configure option as well.

With this, you mean that the option should be configurable per entry? I think it's preferable that the checkbox affects a global setting.

>I would prefer removing the auto-close feature completely as suggested before. It's not time critical to close the window and it removes the double function checkbox that moves around.

The primary motivation for the auto-close was to prevent blocking the auto-close feature since we're using a modal window. I agree it's ugly that the checkbox "moves around" when the text changes. Perhaps this can be fixed, instead of dropping it altogether?

hifi replied:
> With this, you mean that the option should be configurable per entry? I think it's preferable that the checkbox affects a global setting.

I mean that a global option shouldn't be on the dialog.

> Perhaps this can be fixed, instead of dropping it altogether?

Having the auto-close as a global option in the settings and on by default would possibly be a solution. Only show the time on the window then.

adolfogc replied:
@hifi What do you think of the latest changes addressing this issue? Here is a screenshot: 

![screenshot here](https://cdn.pbrd.co/images/H2Z5KS6.png)

hifi commented on src/gui/TotpExportSettingsDialog.cpp:
```diff
@@ -0,0 +1,128 @@
+
+    const QUrl keyUri = totpKeyUri(entry);
+    const QrCode qrc(keyUri.toEncoded());
+    const QVariant maybeIMG = qrc.toQImage(200, 4);
```
Is `QVariant` really necessary here? Can't it just return `null`?

adolfogc replied:
We can use "nullptr," but we would also have to manage memory manually.

droidmonkey replied:
QScopedPointer<...>

adolfogc replied:
I've changed it.

phoerious commented:
Any progress on this? We're feature-freezing 2.3.0 soon.

adolfogc commented:
@phoerious Will try to finish this weekend.

adolfogc commented on src/CMakeLists.txt:
```diff
@@ -252,14 +255,15 @@ target_link_libraries(keepassx_core
+                      ${ZXCVBN_LIBRARIES}
                       ${GCRYPT_LIBRARIES}
                       ${GPGERROR_LIBRARIES}
+ ${qrcode_LIB}
```
Remainder: fix format (change was due to merge resolution after rebasing).

adolfogc commented:
@weslly 

> The app crashes when I try to view the QR code from entries that use the alternative settings format from KeeOtp (key={secret}&size=6&step=30 at the otp attribute)

I'm doing a refactoring of the OTP code in a separate topic branch where I'm addressing this issue. If you're OK with it, I'll work this out in a follow-up PR.

adolfogc commented:
@phoerious In my eyes, this PR is finished. I addressed most of the feedback received in the code review process, with the changes squashed into a single commit; only two improvements are pending. The latter I plan to work out in follow-up PRs if you guys are OK with it.

weslly commented:
@adolfogc Wouldn't be a good idea to merge code that crashes the application in some situations. Can you disable the feature for the problematic items until the issue is fixed in the separate PR?

phoerious commented:
I feel like this needs more testing. I hope you don't mind if I postpone it to 2.4.

hifi commented:
I really like the scalable SVG output! Set the QPen width to zero to make the output cleaner (now 0.1), otherwise it has a border around the rectangles and it causes weird dots in the output when scaling. It also removes spacing between the rectangles caused by the pen.

Also the actual QR code needs to either force the window resize policy to keep aspect ratio or it needs to float inside the window keeping its own aspect ratio only.

phoerious commented:
You need to force a Docker image rebuild by incrementing REBUILD_COUNTER. Otherwise CI will continue to fail.

adolfogc commented:
@phoerious Thanks that fixed the issue I was having with CI. 

@hifi I tried fixing the aspect ratio issue by subclassing `QSvgWidget` and overriding its `hasHeightForWidth/0` and `heightForWidth\1` methods, it didn't work though. Do you have a suggestion on how to fix this?

hifi commented:
@adolfogc I think you need to have a layout that allows space around the widget (with spacers?) so that something will fill the window when it's not a perfect square.

DJCrashdummy commented:
just my 2 cents about the auto-close feature:
- what about a line at the general security-settings in the timeout-section, to uncheck - this should be activated by default - an option to "**simultaneously close the TOTP QR-code window with the database**" an then simply close it a second before the whole database gets locked?! _...or maybe even hardcode this behavior?_
--> this ensures that the auto-close timeouts match (e.g. i configured my database to close after 30sec, so 45sec for the qr-code doesn't make sense).
- and maybe add an additional line similar to the clipboard-setting to optional auto-close the qr-code after a given time - in most cases earlier, which will make the first setting irrelevant, if used.
--> the question is, should there values higher than for the database-close be allowed or not...? - if yes, the first setting resp. its handling is not completely irrelevant.

the first setting (resp. its default behavior) will prevent conflicts or headaches for newbies and placing the setting there will make it easier to understand that the auto-close of the qr-code is due to security reasons.

adolfogc commented:
@DJCrashdummy Currently, you can set this value as shown below:
![screenshot here](https://cdn.pbrd.co/images/H2Z5KS6.png)

>the question is, should there values higher than for the database-close be allowed or not...? - if yes, the first setting resp. its handling is not completely irrelevant.

Though one would expect the timeout used for the DB auto-close to be higher than the one used for closing the QR code window, I guess it does make sense to use the minimum of those two timeout values as you're suggesting. What about letting the user specify the timeout value and default to using `min(dbAutoCloseTimeout, qrCodeWindowAutoCloseTimeout)` for deciding when to close the QR code window?

DJCrashdummy commented:
> I guess it does make sense to use the minimum of those two timeout values, as you're suggesting.

no, i'm rather suggesting an additional checkbox (if not hardcoded?) to ensure, that the qr-code window gets always closed when the database gets locked, whether because of timeout or anything else. so this would also apply to the window minimizing behavior, session-lock, lid-close etc.
**--> IMHO this is an essential security-behavior, because the qr-code contains parts of my database!**
this behavior must overrule the qr-timeout and so may but doesn't have to make the qr-timeout irrelevant...

AFTER this essential feature is implemented (and only if optional as checkbox), and both timeouts are activated, this question you are trying to answer arises.
...but then IMHO it wouldn't be that tragically to allow the qr-window to stay open although the database is already closed.

droidmonkey commented:
@adolfogc can you rebase this on develop? Now that the code format is merged we are ready to merge 2.4 PR's.

adolfogc commented:
@droidmonkey I'll try to work on it this weekend.

droidmonkey commented:
@adolfogc do you think this needs anymore work?

TheZ3ro commented:
Status on this PR?

adolfogc commented:
@droidmonkey, @TheZ3ro: Sorry for the late reply guys; due to work, I haven't had the chance to work on this PR lately. There are still some changes I'd like to make.

droidmonkey commented:
Hello @adolfogc, will you be able to make your changes? I would love to merge this in to develop. If not, can you describe what you want to change and maybe someone else can do it?

adolfogc commented:
Hi @droidmonkey, sorry for the late reply. Thanks for the improvements you made in cdd26c3!  I'm currently unable to work on this PR, so I was wondering if perhaps you can finish it by implementing these changes:
- remove the unnecessary setting and fix the timeout to 45 seconds, as  you suggested 
- close the QR code dialog automatically when the DB gets locked, as suggested by @DJCrashdummy

droidmonkey commented:
This is ready for rebase merge.
